src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+                    case SyntaxKind.ContinueStatement:
+                        {
+                            const label = (<BreakStatement | ContinueStatement>n).label;
+                            if (label) {
+                                if (!contains(seenLabels, label.text)) {
+                                    // attempts to jump to label that is not in range to be extracted
+                                    (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingLabeledBreakOrContinueStatementWithTargetOutsideOfTheRange));
+                                }
+                            }
+                            else {
+                                if (!(permittedJumps & (SyntaxKind.BreakStatement ? PermittedJumps.Break : PermittedJumps.Continue))) {
+                                    // attempt to break or continue in a forbidden context
+                                    (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingConditionalBreakOrContinueStatements));
+                                }
+                            }
+                            break;
+                        }
+                    case SyntaxKind.AwaitExpression:
+                        facts |= RangeFacts.IsAsyncFunction;
+                        break;
+                    case SyntaxKind.YieldExpression:
+                        facts |= RangeFacts.IsGenerator;
+                        break;
+                    case SyntaxKind.ReturnStatement:
+                        if (permittedJumps & PermittedJumps.Return) {
+                            facts |= RangeFacts.HasReturn;
+                        }
+                        else {
+                            (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingConditionalReturnStatement));
+                        }
+                        break;
+                    default:
+                        forEachChild(n, visit);
+                        break;
+                }
+
+                permittedJumps = savedPermittedJumps;
+            }
+        }
+    }
+
+    /**
+     * Computes possible places we could extract the function into. For example,
+     * you may be able to extract into a class method *or* local closure *or* namespace function,
+     * depending on what's in the extracted body.
+     */
+    export function collectEnclosingScopes(range: TargetRange): Scope[] {
+        let current: Node = isArray(range.range) ? firstOrUndefined(range.range) : range.range;
+        if (range.facts & RangeFacts.UsesThis) {
+            // if range uses this as keyword or as type inside the class then it can only be extracted to a method of the containing class
+            const containingClass = getContainingClass(current);
+            if (containingClass) {
+                return [containingClass];
+            }

```

Comment:
What if we use `this` but there is no containing class? Do we still want to offer a refactor?